### PR TITLE
Optimize constant offset loads/stores in translation

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit implements an optimization when lowering Wasm bytecode to CLIF to skip a bounds check when the offset in memory is statically known. This comes up in C/C++/Rust when `static` memory is addressed for example where LLVM emits an `i32.const 0` as the base address and puts the address of the variable in the `offset` instruction field. This isn't necessary for 64-bit platforms but when explicit bounds-checks are required this can help to eliminate a constant-true bounds-check.

### CODE REVIEW

File: tests/disas/pulley/memory-inbounds.wat [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  (memory 1)
+
+  (func (result i32)
+    (i32.const 0)
+    i32.load
+  )
+
+  (func (result i32)
+    (i32.const 100)
+    i32.load
```
Comment (sunfishcode) [inline tests/disas/pulley/memory-inbounds.wat:14]: Could you add a few tests with interesting non-zero offsets as well?

File: tests/disas/pulley/memory-inbounds.wat [hunk 2]
```diff
@@ -0,0 +1,31 @@
+;;! target = "pulley64"
+;;! test = "compile"
+
+(module
+  (memory 1)
```
Comment (sunfishcode) [inline tests/disas/pulley/memory-inbounds.wat:5]: Would it make sense to specify a max, like`(memory 1 2)`, so that this can test for loads that are in-bounds on the minimum but not the maximum?

Comment (fitzgen) [reply tests/disas/pulley/memory-inbounds.wat:5]: > so that this can test for loads that are in-bounds on the minimum but not the maximum?

I think this should be "in bounds of the maximum but not the minimum", but yeah +1

**Review by fitzgen — APPROVED** [review]
LGTM modulo a nitpick below and Dan's comments as well.

File: cranelift/codegen/src/ir/immediates.rs [hunk 3]
```diff
@@ -107,6 +108,25 @@ impl Imm64 {
         let sign_extended = (self.0 << delta) >> delta;
         Imm64(sign_extended)
     }
+
+    /// Zero extend this immediate as if it were a signed integer of the given
+    /// power-of-two width.
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply cranelift/codegen/src/ir/immediates.rs:113]: Oops this is a copy/paste typo
